data/tutorials/language/0lg_08_basic_datatypes.md

```diff
@@ -0,0 +1,686 @@
+    | Druid
+    | Fighter
+    | Monk
+    | Paladin
+    | Ranger
+    | Rogue
+    | Sorcerer
+    | Wizard;;
+type character_class =
+    Barbarian
+  | Bard
+  | Cleric
+  | Druid
+  | Fighter
+  | Monk
+  | Paladin
+  | Ranger
+  | Rogue
+  | Sorcerer
+  | Wizard
+
+# type rectitude = Evil | R_Neutral | Good;;
+type rectitude = Evil | R_Neutral | Good
+
+# type firmness = Chaotic | F_Neutral | Lawful;;
+type firmness = Chaotic | F_Neutral | Lawful
+```
+
+These kinds of variant types can also be used to represent weekdays, cardinal
+directions, or any other fixed-sized set of values that can be given names. A
+total ordering is defined on values following the definition order (e.g., `Druid
+< Ranger`).
+
+Here is how pattern matching can be done on types defined as such.
+```ocaml
+# let rectitude_to_french = function
+    | Evil -> "Mauvais"
+    | R_Neutral -> "Neutre"
+    | Good -> "Bon";;
+val rectitude_to_french : rectitude -> string = <fun>
+```
+
+Note that:
+
+- `unit` is an enumerated-as-variant with a unique constructor: `()`.
+- `bool` is also an enumerated-as-variant with two constructors: `true` and `false`.
+
+A pair `(x, y)` has type `a * b`, where `a` is the type of `x` and `b` is the type of `y`. Some may find it intriguing that `a * b` is called a product. Although this is not a complete explanation, here is a remark which may help in understanding: Consider the product type `character_class * character_alignement`. There are 12 classes and 9 alignments. Any pair of values from those types inhabit the product type. Therefore, in the product type, there are 9 × 12 = 108 values, which also is a product.
+
+#### Constructors With Data
+
+It is possible to wrap data in constructors. The following type has several constructors with data and some without. It represents the different means to refer to a Git [revision](https://git-scm.com/docs/gitrevisions).
+```ocaml
+# type commit =
+  | Hash of string
+  | Tag of string
+  | Branch of string
+  | Head
+  | Fetch_head
+  | Orig_head
+  | Merge_head;;
+type commit =
+    Hash of string
+  | Tag of string
+  | Branch of string
+  | Head
+  | Fetch_head
+  | Orig_head
+  | Merge_head
+
+```
+
+Here is how pattern matching can be used to write a function from `commit` to `string`:
+```ocaml
+# let commit_to_string = function
+  | Hash sha -> sha
+  | Tag name -> name
+  | Branch name -> name
+  | Head -> "HEAD"
+  | Fetch_head -> "FETCH_HEAD"
+  | Orig_head -> "ORIG_HEAD"
+  | Merge_head -> "MERGE_HEAD";;
+val commit_to_string : commit -> string = <fun>
+```
+
+Here, the `function ...` construct is used instead of the `match ... with ...` construct. Previously, example functions had the form `let f x = match x with ...`, and the variable `x` did not appear after any of the `->` symbols. When this is the case, the `function ...` construct can be used instead. It stands for `fun x -> match x with ...` and saves us from finding a name which is used right after and only once.
```

Comment:
```suggestion
Here, the `function ...` construct is used instead of the `match ... with ...` construct. Previously, example functions had the form `let f x = match x with ...`, and the variable `x` did not appear after any of the `->` symbols. When this is the case, the `function ...` construct can be used instead. It stands for `fun x -> match x with ...` and saves us from finding a name which is used immediately after and only once.
```
Suggestion as you sometimes talk about things on the right and left (but not necessary)